### PR TITLE
VMVX mmt4d ukernel

### PR DESCRIPTION
**Pull request by bjacob**
This brings an initial (unoptimized, reference code only) mmt4d ukernel - both `f32f32f32` and `i8i8i32`.

It is covered by the e2e matmul tests: if you purposefully introduce a numerical bug in the ukernel function, `iree_vmvx_mmt4d_f32f32f32` then this test fails: `iree/tests/e2e/matmul/e2e_matmul_mmt4d_f32_small_ukernel_vmvx_local-task` . Ditto for `i8i8i32`.

That the whole reference code is for now in `module.c`, as opposed to being nicely isolated in `iree/builtings/ukernel`, is temporary. I have a few questions to ask about the placeholders in this directory, but it will be so much more concrete to discuss after we are done reviewing this PR so I hope that's OK to split as a separate code move for another PR.

A couple of nontrivial decisions in this PR:

* In `LowerLinalgMicrokernels.cpp` there was a `isUnitInnerStride` helper function. It was only applied to 2D memrefs. The underlying question is how much layout generality do we want ukernels to support, and the existing code embodied a decision on this for 2D arrays, but mmt4d deals with 4D arrays so the question was how to generalize this from 2D to 4D arrays. I chose to generalize `isUnitInnerStride` into `areInnerDimsContiguousRowMajor`. See the comment where it is defined. The lit test, `lower_linalg_microkernels.mlir`, has testcases to cover several edge cases here.

* Similar to what we decided last week for matmul in #10211, there was the question of how to deal with the accumulators that is nonzero in the general case but that we know will often be zero in practice so that we will want to retain the ability to take advantage of that. This is handled here exactly like it was for matmul in #10211. I even reused the flag symbolic constant rather than create a separate one. Yay for weak typing.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp [hunk 1]
```diff
@@ -99,9 +128,11 @@ struct StridedBufferDescriptor {
   Type getElementType() { return memRefType.getElementType(); }
   TypeAttr getElementTypeAttr() { return TypeAttr::get(getElementType()); }
 
-  /// Returns whether the innermost stride is statically 1. Many kernels
-  /// require this, so we provide the convenience here.
-  bool isUnitInnerStride() { return isMemRefUnitInnerStride(memRefType); }
+  // Returns true if all inner dimensions (that is, all but the outer-most dim)
+  // are statically known to be contiguous row-major.
+  bool areInnerDimsContiguousRowMajor() const {
+    return ::mlir::iree_compiler::areInnerDimsContiguousRowMajor(memRefType);
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp:134]: What if we just rename the helper function above to avoid the name aliasing and remove the need for the full namespace qualification. verifyMemRefInnerDimsContiguousRowMajor?

File: compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td [hunk 2]
```diff
@@ -216,6 +216,54 @@ def VMVX_MatmulOp : VMVX_Op<"matmul"> {
...
+    VMVX_Index:$out_offset,
+    VMVX_Index:$out_row_stride,
+
+    // Dimensions.
+    VMVX_Index:$m,
+    VMVX_Index:$n,
+    VMVX_Index:$k,
+
+    // Tile dimensions.
+    VMVX_Index:$m0,
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td:244]: Wdyt of calling these tile_m, tile_n, tile_k?

File: compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td [hunk 3]
```diff
@@ -216,6 +216,54 @@ def VMVX_MatmulOp : VMVX_Op<"matmul"> {
...
+    VMVX_ElementTypeAttr:$out_type,
+    I32Attr:$flags
+  );
+
+  let assemblyFormat = [{
+    `lhs` `` `(` $lhs_buffer `offset` $lhs_offset `row_stride` $lhs_row_stride `:` type($lhs_buffer) `)`
+    `rhs` `` `(` $rhs_buffer `offset` $rhs_offset `row_stride` $rhs_row_stride `:` type($rhs_buffer)`)`
+    `out` `` `(` $out_buffer `offset` $out_offset `row_stride` $out_row_stride `:` type($out_buffer) `)`
+    `mnk` `` `(` $m `,` $n `,` $k `)`
+    `m0n0k0` `` `(` $m0 `,` $n0 `,` $k0 `)`
```
Comment (stellaraccident) [inline compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td:260]: And label this `tile_mnk` if doing the above?

**Review by stellaraccident — APPROVED** [review]
What you propose as answers to your questions makes sense to me. Left a couple of comments about naming that is up to you whether to take them.

File: compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp]: done

File: compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td (hunk 3 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/VMVX/IR/VMVXOps.td]: done
